### PR TITLE
New Spring Boot Security blog post

### PR DESCRIPTION
**Pull request by brunokrebs**

### CODE REVIEW

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+
+Our custom authentication filter overwrites two methods of the base class:
+
+- `attemptAuthentication`: where we parse the user's credentials and issue them to the [`AuthenticationManager`](https://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/authentication/AuthenticationManager.html).
+- `successfulAuthentication`: which is the method called when a user successfully logs in. We use this method to generate a *JWT* to this user.
+
+Our *IDE* will probably complain about the code in this class for two reasons. First, because the code imports four constants from a class that we haven't created yet, `SecurityConstants`. Second, because this class generates *JWTs* with the help of a class called `Jwts`, which belongs to a library that we haven't added as dependency to our project.
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: Don't italicize IDE or JWTs.

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+	}
+}
+```
+
+Note that the authentication filter that we created extends the [`UsernamePasswordAuthenticationFilter` class](http://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/web/authentication/UsernamePasswordAuthenticationFilter.html). When we add a new filter to *Spring Security*, we can explicitly define where in the *filter chain* we want that filter, or we can let the framework figure it out by itself. By extending the filter provided within the security framework, *Spring* can automatically identify the best place to put it in the security chain.
+
+Our custom authentication filter overwrites two methods of the base class:
+
+- `attemptAuthentication`: where we parse the user's credentials and issue them to the [`AuthenticationManager`](https://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/authentication/AuthenticationManager.html).
+- `successfulAuthentication`: which is the method called when a user successfully logs in. We use this method to generate a *JWT* to this user.
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: Don't italicize JWT

`generate a JWT for this user` OR `issue a JWT to this user`

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+public class SecurityConstants {
+	public static final String SECRET = "SecretKeyToGenJWTs";
+	public static final long EXPIRATION_TIME = 864_000_000; // 10 days
+	public static final String TOKEN_PREFIX = "Bearer ";
+	public static final String HEADER_STRING = "Authorization";
+	public static final String SIGN_UP_URL = "/users/sign-up";
+}
+```
+
+This class contains all four constants referenced by the `JWTAuthenticationFilter` class, alongside with a `SIGN_UP_URL` constant that will be used later.
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: `alongside with` -> should be "along with" OR "alongside"

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+				return new UsernamePasswordAuthenticationToken(user, null, new ArrayList<>());
+			}
+			return null;
+		}
+		return null;
+	}
+}
+```
+
+We have extended the [`BasicAuthenticationFilter`](https://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/web/authentication/www/BasicAuthenticationFilter.html) to make *Spring* replace it in the *filter chain* with our custom implementation. The most important part of the filter that we've implemented is the private `getAuthentication` method. This method reads the *JWT* from the `Authorization` header, and then uses [`Jwts`](https://github.com/jwtk/jjwt/blob/master/src/main/java/io/jsonwebtoken/Jwts.java) to validate the token. If everything is in place, we set the user in the [`SecurityContext`](https://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/core/context/SecurityContext.html) and allow the request to move on.
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: Unnecessary italics.

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+	@Override
+	public void configure(AuthenticationManagerBuilder auth) throws Exception {
+		auth.userDetailsService(userDetailsService).passwordEncoder(bCryptPasswordEncoder);
+	}
+}
+```
+
+We have annotated this class with `@EnableWebSecurity` and made it extend `WebSecurityConfigurerAdapter` to take advantage of the default [web security configuration provided by *Spring Security*](https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-security.html). This allows us to fine-tune the framework to our needs by overwriting two methods:
+
+- `configure(HttpSecurity http)`: a method where we can define which resources are public and which are secured—in our case we set the `SIGN_UP_URL` endpoint as being public and everything else as being secured—and also configure custom security filter in the *Spring Security filter chain*.
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: This sentence is long and awkward with the em dashes. I'd recommend splitting it into two sentences instead.

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+- `configure(AuthenticationManagerBuilder auth)`: a method where we defined a custom implementation of [`UserDetailsService`](https://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/core/userdetails/UserDetailsService.html) to load user-specific data in the security framework. We have also used this method to set the encrypt method used by our application (`BCryptPasswordEncoder`).
+
+*Spring Security* doesn't comes with a concrete implementation of `UserDetailsService` that we could use out of the box with our in-memory database. Therefore, we create a new class called `UserDetailsServiceImpl` in the `com.auth0.samples.authapi.user` package to provide one:
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: No italics.

`doesn't comes` -> no "s" on "come"

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+		ApplicationUser applicationUser = applicationUserRepository.findByUsername(username);
+		if (applicationUser == null) {
+			throw new UsernameNotFoundException(username);
+		}
+		return new User(applicationUser.getUsername(), applicationUser.getPassword(), emptyList());
+	}
+}
+```
+
+The only method that we had to implement is the `loadUserByUsername` one. When a user tries to authenticate, this method receives its username, searches in the database for a record containing it, and (if found) returns an instance of `User`. The properties of this instance (`username` and `password`) are then checked against the credentials passed by the user in the login request. This last process is executed outside this class, by the *Spring Security* framework.
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: `the loadUserByUsername one` -> delete "the" and "one".

`receives its username` -> shouldn't refer to users as "it" ;) Change this to "receives the username"

`searches in the database` -> delete "in"

No italics on Spring Security.

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 8]
```diff
@@ -0,0 +1,532 @@
...
+
+We can now rest assured that our endpoints won't be publicly exposed and that we can support authentication and authorization with *JWTs* on *Spring Boot* properly. To check everything, let's run our application (through the *IDE* or through `gradle bootRun`) and issue the following requests:
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: Remove italics from JWTs, Spring Boot, and IDE.

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+}'  http://localhost:8080/tasks
+
+# issue a new GET request, passing the JWT
+# remember to replace xxx.yyy.zzz with the JWT retrieved above
+curl -H "Authorization: Bearer xxx.yyy.zzz" http://localhost:8080/tasks
+```
+
+## Securing Spring Applications with Auth0
+
+Securing *Spring* applications with Auth0 is very easy and brings a lot of great features to the table. With Auth0, we have to write just a few lines of code to get a solid [identity management solution](https://auth0.com/docs/identityproviders), including [single sign-on](https://auth0.com/docs/sso/single-sign-on), [user management](https://auth0.com/docs/user-profile), support for [social identity providers (like Facebook, GitHub, Twitter, etc.)](https://auth0.com/docs/identityproviders), [enterprise (Active Directory, LDAP, SAML, etc.)](https://auth0.com/enterprise), and our [own database of users](https://auth0.com/docs/connections/database/mysql).
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: No italics.

`we have to write just a few lines of code` -> "we only have to write a few lines of code"

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 10]
```diff
@@ -0,0 +1,532 @@
...
+
+[To learn the best way to secure *Spring Security API endpoints* with Auth0, take a look at this tutorial](https://auth0.com/docs/quickstart/backend/java-spring-security). Besides providing tutorials for backend technologies (like *Spring*), [the *Auth0 Docs* webpage also provides tutorials for *Mobile/Native apps* and *Single-Page applications*](https://auth0.com/docs).
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: Too many italics.

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 11]
```diff
@@ -0,0 +1,532 @@
...
+
+## Conclusion
+
+Securing *RESTful Spring Boot API* with *JWTs* is not a hard task. This article showed that by creating a couple of classes, and extending a few others provided by *Spring Security*, we can protect our endpoints from unknown users, enable users to register themselves, and authenticate existing users based on *JWTs*.
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: Remove italics.

`classes, and extending` -> delete comma

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 12]
```diff
@@ -0,0 +1,532 @@
...
+
+Of course that for a production-ready application we would need a few more features, like password retrieval, but the article demystified the most sensible parts of dealing with *JWTs* to authorize requests on *Spring Boot* applications.
```
Comment (kmaida) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: Remove italics.

File: _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown [hunk 13]
```diff
@@ -0,0 +1,532 @@
...
+During the authentication process, when a user successfully logs in using their credentials, a *JSON Web Token* is returned and must be saved locally (typically in local storage). Whenever the user wants to access a protected route or resource (an endpoint), the user agent must send the *JWT*, usually in the `Authorization` header using the [*Bearer schema*](http://self-issued.info/docs/draft-ietf-oauth-v2-bearer.html), alongside with the request.
+
+When a backend server receives a request with a *JWT*, the first thing to do is to validate the token. This consists of a series of steps, and if any of these fails then the request must be rejected. The following list shows the validation steps needed:
+
+- Check that the JWT is well formed
+- Check the signature
+- Validate the standard claims
+- Check the Client permissions (scopes)
+
+We won't get into the nitty-gritty details about *JWTs* in this article but, if needed, [this resource can provide more about information about *JWTs*](https://auth0.com/docs/jwt) and this [resource about *JWT* validation](https://auth0.com/docs/api-auth/tutorials/verify-access-token).
```
Comment (brunokrebs) [inline _posts/2017-07-13-developing-secure-restful-apis-with-spring-boot.markdown]: Well, I thought I should be consistent throughout the article.

**Pull request closed without merging**